chore(flags): Use PostgeSQL's session-level statement_timeout

haacked opened the pull request:
## Problem

Our queries don't have any timeouts specified. By default that means a query can run as long as it takes. Without timeouts, slow queries can pile up and consume database connections, potentially bringing down the entire service.
 
During high loads, timeouts help by:

1. Preventing resource exhaustion - releasing database connections that would otherwise be tied up in slow/hung queries
2. Avoiding cascading failures - stopping the pile-up of requests that would eventually crash the system
3. Making failure observable - you get timeout errors instead of silent hangs, so you know there's a problem

## Changes

Ensure that we're using aggressive, but reasonable timeouts in production. @andyzzhao, would love your 👀 on the defaults I chose.

Also removed the `run_query` method on `Client` because it was only used by unit tests and caused a lot of confusion. Now, unit tests query the database the same way production code does, albeit with different timeouts.

Did not add `tokio::timeout` (aka a client timeout) this time around. We didn't have them before and adding them would change every call-site. I'd want to find an approach that is more centralized if we decide to add these. And we get most of the benefit from having PostgreSQL timeouts.

## How did you test this code?

Unit tests
Manual testing: Hit `/flags` service. Made sure writing to `posthog_featureflaghashkeyoverride` still works.

<!-- Docs reminder: If this change requires updated docs, please do that! Engineers are the primary people responsible for their documentation. 🙌 -->

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

haacked committed:
Use PostgeSQL's session-level statement_timeout

Ensure that we're using aggressive, but reasonable timeouts in production. Prior to this change, we didn't have any timeouts configured for PostgreSQL.

Also removed the `run_query` method on `Client` because it was only used by unit tests. Now, unit tests query the database the same way production code does, albeit with different timeouts.

Did not add `tokio::timeout` (aka a client timeout) this time around. We didn't have them before and adding them would change every call-site. I'd want to find an approach that is more centralized. And we get most of the benefit from having PostgreSQL timeouts.

andyzzhao commented on rust/common/database/src/lib.rs:
```diff
@@ -111,6 +156,39 @@ pub fn is_foreign_key_constraint_error(error: &SqlxError) -> bool {
 }
 
+/// Determines if a sqlx::Error represents a timeout-related failure
+pub fn is_timeout_error(error: &SqlxError) -> bool {
```
Tuning the timeouts will depend on what behavior we see in prod so it might be worth adding some logs or metrics on which error was seen

haacked commented on rust/common/database/src/lib.rs:
```diff
@@ -4,20 +4,27 @@ use anyhow::Result;
+// Default database timeouts - optimized for production low-latency flag evaluation
+pub const DEFAULT_TIMEOUTS: DatabaseTimeouts = DatabaseTimeouts {
+    statement_timeout: Duration::from_millis(300), // Aggressive for fast flag evaluation
+    lock_timeout: Duration::from_millis(100),      // Avoid blocking on locks
```
Yeah, I just recently learned about this: https://www.postgresql.org/docs/current/runtime-config-client.html

haacked committed:
Add granular timeout types to error metrics

Extend TimeoutError to include specific timeout types (query_canceled,
lock_not_available, pool_timeout, etc.) and return timeout:type
prometheus labels for better observability.

haacked committed:
Fix Rust compilation errors and warnings

- Fix TimeoutError tuple variant pattern matching in status code mapping
- Fix TimeoutError constructor in test assertions
- Add explicit lifetime annotation to fix lifetime elision warning
- All 407 tests now passing after database migration setup

The TimeoutError variant is defined as TimeoutError(Option<String>) but was
being used as a unit variant in pattern matching and tests.
